File: customGTApp/src/main/java/com/customGTApp/service/ProductService.java
package com.customGTApp.service;

import com.customGTApp.model.Photo;
import com.customGTApp.model.Product;

import java.util.List;

/**
 * The contract our product service need to implement so the application works correctly and the code is more organized.
 * This interface will be implemented by ProductServiceImpl and other classes.
 */

public interface ProductService {
    /**
     * Method for returning all products.
     * @return a list of all products from the DataBase.
     */
    List<Product> findAllProducts();

    /**
     * Method for finding a product by id
     * @param id the id of the product we need to find
     * @return the product found
     */
    Product findProductById(Long id);

    /**
     * Method for adding a product
     * @param p the product we want to add to DataBase
     * @return the product added
     */
    Product addProduct(Product p);

    /**
     * Method for updating a product
     * @param p the product with the fields updated
     * @return the updated product
     */
    Product updateProduct(Product p);

    /**
     * Method for deleting a product by id
     * @param id the id of the product we want to delete
     */
    void deleteProductById(Long id);

    /**
     * Method to list all the photos of a product
     * @param productId the id of a product we want to list his photos
     * @return the photos list of the product
     */
    List<Photo> findPhotosByProductId(Long productId);

    /**
     * Method to update the quantity of a product
     * @param productId the id of the product we want to update the quantity
     * @param quantity the new quantity of the product
     * @return the product with the updated quantity
     */
    Product updateQuantity(Long productId, int quantity);

    /**
     * Method to update the price of a product
     * @param productId the id of the product we want to update the price
     * @param price the new price of the product
     * @return the product with the updated price
     */
    Product updatePrice(Long productId, float price);

}

File: customGTAppFront/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { Product } from '../models/product';
import { Observable } from 'rxjs';
import { Photo } from '../models/photo';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private apiUrl = environment.apiUrl + '/product';

  constructor(private http: HttpClient) { }

  public getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl + '/all');
  }

  public getAllPhotosByProductId(productId: number): Observable<Photo[]> {
    return this.http.get<Photo[]>(this.apiUrl + '/' + productId + '/photos');
  }

  public getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(this.apiUrl + '/find/' + id);
  }

  public addProduct(product: Product): Observable<Product> {
    return this.http.post<Product>(this.apiUrl + '/add', product);
  }

  public updateProduct(product: Product): Observable<Product> {
    return this.http.put<Product>(this.apiUrl + '/update', product);
  }

  public deleteProductById(id: number): Observable<Product> {
    return this.http.delete<Product>(this.apiUrl + '/delete/' + id);
  }

  public updateQuantity(productId: number, quantity: number): Observable<Product> {
    const url = `${this.apiUrl}/updateQuantity/${productId}?quantity=${quantity}`;
    return this.http.put<Product>(url, null);
  }

  public updatePrice(productId: number, price: number): Observable<Product> {
    const url = `${this.apiUrl}/updatePrice/${productId}?price=${price}`;
    return this.http.put<Product>(url, null);
  }

}
